heron/common/src/java/com/twitter/heron/common/network/HeronServerTester.java

```diff
@@ -0,0 +1,367 @@
+   */
+  public HeronServerTester(HeronServer server,
+                           TestRequestHandler requestHandler,
+                           TestResponseHandler responseHandler,
+                           Duration responseReceivedTimeout) throws IOException {
+    this(server);
+    this.responseReceivedSignal = new CountDownLatch(1);
+    this.responseReceivedTimeout = responseReceivedTimeout;
+    this.client = new TestClient(new NIOLooper(),
+        server.getEndpoint().getHostName(), server.getEndpoint().getPort(),
+        responseReceivedSignal, requestHandler, responseHandler);
+  }
+
+  /**
+   * Constructor where a custom client is to be used. The start() method starts both the server and
+   * the client.
+   * @param server server to test
+   * @param client client to test
+   */
+  public HeronServerTester(HeronServer server, HeronClient client) {
+    this(server);
+    this.client = client;
+  }
+
+  private HeronServerTester(HeronServer server) {
+    this.server = server;
+    this.threadsPool = Executors.newFixedThreadPool(2);
+    this.serverStartedSignal = new CountDownLatch(1);
+  }
+
+  public void start() throws InterruptedException {
+    // First run Server
+    runServer();
+
+    // Then run Client
+    runClient();
+
+    // Wait to make sure message was sent and response was received
+    if (responseReceivedTimeout != null) {
+      await(responseReceivedSignal, responseReceivedTimeout);
+    }
+  }
+
+  public void stop() {
+    threadsPool.shutdownNow();
+
+    server.stop();
+
+    if (client != null) {
+      client.stop();
+      client.getNIOLooper().exitLoop();
+    }
+
+    server.getNIOLooper().exitLoop();
+  }
+
+  /**
+   * Convenience method to wait on a latch with a default timeout. If the timeout is reached, the
+   * fail() method will be invoked.
+   */
+  public static void await(CountDownLatch latch) {
+    await(latch, DEFAULT_LATCH_TIMEOUT);
+  }
+
+  /**
+   * Convenience method to wait on a latch with a default timeout. If the timeout is reached, the
+   * fail() method will be invoked.
+   */
+  public static void await(CountDownLatch latch, Duration timeout) {
+    try {
+      latch.await(timeout.toMillis(), TimeUnit.MILLISECONDS);
+    } catch (InterruptedException e) {
+      fail(String.format(
+          "Await latch failed to release until timeout of %s was reached. Check latch logic.",
+          timeout));
+    }
+  }
+
+  /**
+   * Wait for a multiCountMetric to reach or exceed a specific value before returning. Assumes that
+   * values are being increased over time by the implementor.
+   * @param multiCountMetric the multiCountMetric to check
+   * @param key the key of the value to be checked
+   * @param expected the minimum expected value to observe before returning
+   * @param timeout how long to wait for the expected value before failing
+   * @throws InterruptedException
+   */
+  public static void awaitMultiCountMetric(MultiCountMetric multiCountMetric, String key,
```

Comment:
This method seems not generic for most test cases to utilize. Can we put it back to the MetricsManagerServerTest?